### PR TITLE
[data-logging] AsyncLogger implementation

### PR DESCRIPTION
**Pull request by bfineran**
basic implementation for a `AsyncLogger` that submits logging jobs that do not block the main process.

current implementation uses a threadpool, however should further evaluate tradeoffs with process pool, and potential future need for a more sophisticated job worker

open questions:
* how to best propagate errors after jobs have been submitted and user requests have been returned. currently issuing a python `error` log that displays the error message however this is more of a "silent" failure which is not ideal - a separate job worker may be able to better handle this
* exploration of limitations of executors and potential overhead added

**test_plan:**
included tests to verify that calls are non blocking, numpy arrays and functions may be submitted, verification of job run and error propagation

more thorough testing should be added in E2E tests

### CODE REVIEW

File: src/deepsparse/loggers/__init__.py [hunk 1]
```diff
@@ -12,10 +12,15 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from .base_logger import *
-
 # flake8: noqa
+# isort: skip_file
```
Comment (dbogunowicz) [inline src/deepsparse/loggers/__init__.py:16]: to disable isort from enforcing his own, messy line ordering?

File: src/deepsparse/loggers/async_logger.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    def log(self, identifier: str, value: Any, category: MetricCategories):
+        """
+        Forward log calls to wrapped logger to run asynchronously
+
+        :param identifier: The name of the item that is being logged.
+        :param value: The data structure that the logger is logging
+        :param category: The metric category that the log belongs to
+        """
+        job_future = self._job_pool.submit(
```
Comment (dbogunowicz) [inline src/deepsparse/loggers/async_logger.py:54]: Nice, also a very well-documented set of tests.

I think optimizing the `AsyncLogger` at this point would be less important than optimizing the pre/post-processing of the pipelines. I would for now leave this PR as it is. We should probably talk to the runtime team/use the cprofiler with snakeviz to understand the possible overhead we are adding with the data logging. But I do not see this as the prority atm.

**Review by dbogunowicz — APPROVED** [review]

File: src/deepsparse/loggers/async_logger.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    the main process. No call back/returned future currently provided
+
+    :param logger: logger object to wrap
+    :param max_workers: maximum logging tasks to run in the job pool at once.
+        defaults to 1
+    """
+
+    def __init__(self, logger: BaseLogger, max_workers: int = 1):
+        self.logger = logger
+        self._job_pool: Executor = ThreadPoolExecutor(max_workers=max_workers)
```
Comment (corey-nm) [inline src/deepsparse/loggers/async_logger.py:44]: Option to pass this in to this object? I think server could pass it the thread pool executor it owns to this so they share one. Or do we explicitly want a separate one I guess? How do 2 thread pool executors interact?

File: src/deepsparse/server/build_logger.py [hunk 4]
```diff
@@ -58,7 +59,10 @@ def build_logger(server_config: ServerConfig) -> BaseLogger:
 
     leaf_loggers = build_leaf_loggers(loggers_config)
     loggers = build_function_loggers(server_config.endpoints, leaf_loggers)
-    return MultiLogger(loggers)
+    return AsyncLogger(
```
Comment (corey-nm) [inline src/deepsparse/server/build_logger.py:62]: 🔥

File: tests/deepsparse/loggers/helpers.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+from deepsparse.loggers import BaseLogger, MetricCategories
+
+
+__all__ = [
+    "ErrorLogger",
+    "FileLogger",
+    "NullLogger",
+    "SleepLogger",
+]
```
Comment (corey-nm) [inline tests/deepsparse/loggers/helpers.py:32]: i love em all

**Review by corey-nm — APPROVED** [review]
lgtm

File: src/deepsparse/loggers/async_logger.py (hunk 3 above)
Comment (bfineran) [reply src/deepsparse/loggers/async_logger.py:44]: yes I like this, let's get it added in later. 2 thread pools should be ok depending on usage, can lead to deadlock in certain scenarios. let's come back to this in testing

**Pull request merged**